fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/KeyExpressionComparisons.java

```diff
@@ -341,6 +373,96 @@ public void addToScanComparisonsBuilder(@Nonnull ScanComparisons.Builder builder
+                // (2) We cannot match to a partially matched part of a Then
+                // Instead, find _only_ the first unmatched child and check if it matches. Otherwise, there's
+                // nothing here.
+                final List<KeyExpressionWithComparison> matchedChildren = new ArrayList<>();
+                final Iterator<KeyExpressionWithComparison> childrenIterator = children.iterator();
+                while (childrenIterator.hasNext()) {
+                    final KeyExpressionWithComparison child = childrenIterator.next();
+                    if (child.getMatchedComparisonType().equals(MatchedComparisonType.NOT_MATCHED)) {
+                        // This is the first unmatched child of the ThenKeyExpression. This one must match or we don't
+                        // have a match at all.
+                        final NestedAndUnnestingFunction nestedChild = child.asNestedWith(nestedContext);
+                        if (nestedChild == null) {
+                            return null;
+                        }
+                        final List<KeyExpressionWithComparison> remainingChildren = new ArrayList<>();
+                        childrenIterator.forEachRemaining(remainingChildren::add);
+                        return nestedChild.combineUnnesting(result -> {
+                            final List<KeyExpressionWithComparison> allChildren = new ArrayList<>(children.size());
+                            allChildren.addAll(matchedChildren);
+                            allChildren.add(result);
+                            allChildren.addAll(remainingChildren);
+                            return withChildren(allChildren);
+                        });
+                    } else if (child.getMatchedComparisonType().equals(MatchedComparisonType.MATCHED)) {
+                        matchedChildren.add(child);
+                    } else {
+                        break;
+                    }
+                }
+                return null;
+            } else {
+                // The parent field is not repeated, so the view should include all parts of the then that have this
+                // nesting.
+                final List<KeyExpressionWithComparison> childrenWithMatchingNesting = new ArrayList<>();
+                final List<UnnestingFunction> unnestingFunctions = new ArrayList<>();
+                for (KeyExpressionWithComparison child : children) {
+                    final NestedAndUnnestingFunction nestedChild = child.asNestedWith(nestedContext);
+                    if (nestedChild != null) {
+                        final int nestedChildIndex = childrenWithMatchingNesting.size();
+                        childrenWithMatchingNesting.add(nestedChild.getNested());
+                        unnestingFunctions.add(result -> nestedChild.getUnnestingFunction()
+                                .unnest(result.children.get(nestedChildIndex)));
```

Comment:
Trivially, it would appear that `result.children.get(nestedChildIndex) == nestedChild.getNested()`, but I assume that this can change due to rewriting things in the exploration stage?